examples/files/main.go

```diff
@@ -0,0 +1,280 @@
+package main
+
+import (
+	"bufio"
+	"database/sql"
+	"fmt"
+	"io"
+	olog "log"
+	"net/http"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"syscall"
+
+	"github.com/kelseyhightower/envconfig"
+
+	"nimona.io/internal/net"
+	"nimona.io/internal/version"
+	"nimona.io/pkg/blob"
+	"nimona.io/pkg/context"
+	"nimona.io/pkg/crypto"
+	"nimona.io/pkg/localpeer"
+	"nimona.io/pkg/log"
+	"nimona.io/pkg/network"
+	"nimona.io/pkg/object"
+	"nimona.io/pkg/objectmanager"
+	"nimona.io/pkg/objectstore"
+	"nimona.io/pkg/peer"
+	"nimona.io/pkg/resolver"
+	"nimona.io/pkg/sqlobjectstore"
+
+	_ "github.com/arl/statsviz"
+)
+
+type fileTransfer struct {
+	local         localpeer.LocalPeer
+	objectmanager objectmanager.ObjectManager
+	objectstore   objectstore.Store
+	resolver      resolver.Resolver
+	listener      net.Listener
+	config        *config
+}
+
+// nolint: lll
+type config struct {
+	Peer struct {
+		PrivateKey  crypto.PrivateKey `envconfig:"PRIVATE_KEY"`
+		BindAddress string            `envconfig:"BIND_ADDRESS" default:"0.0.0.0:0"`
+		Bootstraps  []peer.Shorthand  `envconfig:"BOOTSTRAPS"`
+	} `envconfig:"PEER"`
+
+	ReceivedFolder string `envconfig:"RECEIVED_FOLDER" default:"received_files"`
+	Debug          struct {
+		MetricsPort string `envconfig:"METRICS_PORT"`
+	} `envconfig:"DEBUG"`
+}
+
+func main() {
+	args := os.Args
+	if len(args) != 3 {
+		fmt.Printf("usage: %s <get/serve> <hash/file>\n", args[0])
+		return
+	}
+
+	ctx := context.New(
+		context.WithCorrelationID("nimona"),
+	)
+	logger := log.FromContext(ctx).With(
+		log.String("build.version", version.Version),
+		log.String("build.commit", version.Commit),
+		log.String("build.timestamp", version.Date),
+	)
+
+	cfg := &config{}
+	if err := envconfig.Process("nimona", cfg); err != nil {
+		logger.Fatal("error processing config", log.Error(err))
+	}
+
+	go func() {
+		if cfg.Debug.MetricsPort == "" {
+			return
+		}
+
+		olog.Println(
+			http.ListenAndServe(
+				fmt.Sprintf("localhost:%s", cfg.Debug.MetricsPort),
+				nil,
+			))
+	}()
+
+	ft, err := newFileTransfer(ctx, cfg, logger)
+	if err != nil {
+		logger.Fatal("error initializing", log.Error(err))
+	}
+	defer ft.close()
+
+	command := args[1]
+	param := args[2]
+
+	switch command {
+	case "get":
+		ft.get(ctx, object.Hash(param), logger)
+	case "serve":
+		ft.serve(ctx, param, logger)
+	default:
+		fmt.Println("command not supported")
+		return
+	}
+
+	// register for termination signals
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	// and wait for one
+	<-sigs
+
+}
+
+func (ft *fileTransfer) serve(
+	ctx context.Context,
+	filename string,
+	logger log.Logger,
+) {
+
+	f, err := os.Open(filename)
+	if err != nil {
+		logger.Fatal("failed to open file", log.Error(err))
```

Comment:
Should we make the things the user needs to see `fmt.Println`s instead of JSON logs?